Objects/unicodeobject.c

```diff
@@ -1429,12 +1429,12 @@ _copy_characters(PyObject *to, Py_ssize_t to_start,
     assert(PyUnicode_Check(from));
     assert(from_start + how_many <= PyUnicode_GET_LENGTH(from));
 
-    assert(PyUnicode_Check(to));
-    assert(to_start + how_many <= PyUnicode_GET_LENGTH(to));
-
     if (how_many == 0)
         return 0;
 
+    assert(PyUnicode_Check(to));
```

Comment:
This assertion can be left above. If this assertion crashes, I think there is an issue with the caller. Which assertion is actually crashing? is it `PyUnicode_Check` or is it the other one?

If `to` is actually NULL, we should instead replace this assertion by `assert((to == NULL && how_many == 0)|| PyUnicode_Check(to))` or fix the calling functions.

Reply:
> This assertion can be left above. If this assertion crashes, I think there is an issue with the caller. Which assertion is actually crashing? is it `PyUnicode_Check` or is it the other one?

It is `PyUnicode_Check`.


> If `to` is actually NULL, we should instead replace this assertion by `assert((to == NULL && how_many == 0)|| PyUnicode_Check(to))`

Agreed.

Reply:
For `PyUnicode_GET_LENGTH` also add `(to == NULL && how_many == 0)||`

`PyUnicode_GET_LENGTH` -> `_PyASCIIObject_CAST` ->  `PyUnicode_Check`
https://github.com/python/cpython/blob/b9a492b809d8765ee365a5dd3c6ba4e5130a80af/Include/cpython/unicodeobject.h#L175

Perhaps it's enough to just leave only this assert
```
assert((to == NULL && how_many == 0) || (to_start + how_many <= PyUnicode_GET_LENGTH(to)));
```

Reply:
Up to you. Assertions can be non-optimal and we tend to be as clear as possible. Alternatively, you can just add `assert(to == NULL || PyUnicode_Check(to));`. That way, we see that you're testing for the "type", while in the second assertion (with `how_many`) you're testing for the values consistency.